### PR TITLE
Implement proper quoting/escaping for filenames

### PR DESCRIPTION
**Pull request by blyxxyz**
The GNU coreutils escape special characters in filenames and adjust quoting depending on the content:
```
$ rmdir file
rmdir: failed to remove 'file': No such file or directory
$ rmdir "can't"
rmdir: failed to remove "can't": No such file or directory
$ rmdir $'foo\n\xFF'
rmdir: failed to remove 'foo'$'\n\377': No such file or directory
```
This PR implements similar handling for uutils:
```
$ uu_rmdir file
uu_rmdir: failed to remove 'file': No such file or directory
$ uu_rmdir "can't"
uu_rmdir: failed to remove "can't": No such file or directory
$ uu_rmdir $'foo\n\xFF'
uu_rmdir: failed to remove $'foo\n\xFF': No such file or directory
```
It's not identical, but like the coreutils output it can be pasted into many shells (like [bash](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html) but also others).

On Windows PowerShell's syntax is used instead.

An extension trait adds a method `quote()` to everything that implements `AsRef<OsStr>`, so not just `&Path` but also `&str`. That way it can be used almost as a drop-in replacement for `path.display()` (you still have to remove the quotes from the format string) and also in places where ordinary strings are used.

So far I've only used it in `rmdir`, as a proof of concept. Grepping for `'{}'` turns up a lot of other places where it would be appropriate.

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
First of all, very good catch! It hadn't even occurred to me yet that filesnames in error messages need quoting.

It's a little bit unfortunate that you've done this from scratch, as there is already a quoting implementation for `ls` that could be reused: https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/quoting_style.rs. 

I believe that implementation matches the GNU examples you mentioned exactly. However, there is a lot of nice things to this PR and I think the best way forward is to use the quoting from `ls`, but your API.

Specifically, I would suggest:
- Keep the escaping of `ls`, but move it into `uucore`.
- Create a wrapper around that escaping with your `Quotable` trait (I think `QuotingStyle::Shell { escape: true, always_quote: true, show_control: false }` would be the right setting to use)
- Add the PowerShell quoting to the `ls` escaping for Windows.

Does that sound reasonable? Let me know if you have any questions about the code in `ls`

**Comment by blyxxyz** [comment]
I deliberately deviated from GNU's quoting style because I don't think it's very readable:
- Displaying bytes in octal is a very uncommon piece of syntax (`\xNN` is broadly supported and less ambiguous)
- It likes to unnecessarily switch between quoting styles instead of just putting everything between `$''` when necessary
- It likes to start off with a redundant `''`

I think it's less important to be byte-for-byte identical than to be able to be *used* interchangeably, and this achieves that. As far as I could find, any shell that supports the syntax GNU outputs also supports the syntax this implementation outputs.

I also notice a few problems with the current `ls` quoting:
- Invalid unicode isn't preserved
- `\`, `$` and ``​`​`` between double quotes aren't escaped (`can'$t` renders as `"can'$t"`)

(While experimenting I also discovered that GNU escapes codepoints that haven't been assigned yet. I'm not sure if that would be appropriate for us, as unicode information is hardcoded into the binary while GNU presumably gets it from the system.)

But it does have some tricks this one doesn't, this one doesn't support unquoted output. Maybe the implementations can be merged?

**Comment by tertsdiepraam** [comment]
> I deliberately deviated from GNU's quoting style because I don't think it's very readable

I agree, GNU-style doesn't look too good.

> But it does have some tricks this one doesn't, this one doesn't support unquoted output. Maybe the implementations can be merged?

I think it'd be pretty cool if you could add your quoting styles to the `ls` (call them `QuotingStyle::UShell` and `QuotingStyle::PowerShell`, or something like that). They would need to be rewritten to match the `Iterator`-based approach (or all the styles in `ls` would need to be adapted to your approach)

> Invalid unicode isn't preserved

Right, it should use bytes instead of chars. I'll try to change that soon.

> `\`, `$` and ​`` ` ``​ between double quotes aren't escaped (can'$t renders as "can'$t")

Interesting! The fix is easy, I'll make a quick PR.

**Comment by blyxxyz** [comment]
> Right, it should use bytes instead of chars

It's trickier: sometimes you need `char`, sometimes `u8`, sometimes `u16` (on Windows). I'll try to fit things together tomorrow and add it to this PR.

**Comment by blyxxyz** [comment]
@tertsdiepraam I've made most of the shell quoting style defer to `uucore::display`. I think that the other styles are ultimately too niche to pull into `display`, so it's better to keep them in `ls`. Can you take a look to see if you agree with how it works?

I think this obsoletes #2622.

The other code in `display` should also be interesting to read if you want to start preserving invalid unicode in the other styles. On Unix you can use `display::from_utf8_iter` to read `char`s mixed with raw bytes (currently private, feel free to make public), and on Windows you can use `std::char::decode_utf16` to mix `char`s with `u16`s. There's no cross-platform solution, unfortunately. For the `c` format Windows's [`wchar` literals](https://docs.microsoft.com/en-us/cpp/cpp/string-and-character-literals-cpp?view=msvc-160) might be interesting.

**Comment by tertsdiepraam** [comment]
This is not quite what I had in mind (which is my fault for not being clear), but I do really appreciate the effort you put in, especially your research and experimentation about what characters should be escaped and how.

I agree that some parts of the `ls` code are too niche for `uucore`, so merging everything would indeed be weird. Therefore, I actually prefer to keep them separate (basically roll back 2 commits). But that might just be because I'm biased towards the `ls` code, so I'll defer the decision to merge this to other maintainers (@miDeb , maybe?).

**Comment by blyxxyz** [comment]
> This is not quite what I had in mind

I realized, but I didn't see a good way to turn it into different styles in `quoting_style`. They're platform-specific, the Unix escaping doesn't compile on Windows and vice-versa. Malformed UTF-8 and malformed UTF-16 aren't interchangeable.

This way seemed an ok compromise because it kept both sides mostly intact, just linked them together.

It also fixed a minor bug, see [here](https://github.com/uutils/coreutils/pull/2621/commits/55a1259206181b0e32d66b75642e103394aa3bba#diff-bca01c3cc10029bf506ed91d70bb32d74f702e708f5df69d2dff19dcdc90dc3dR414). (GNU ls adds quotes in that case.)

Are your concerns about behavior or about structure?

File: src/uucore/src/lib/mods/display.rs [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+impl_as_ref!(std::path::Iter<'_>);
+impl_as_ref!(&'_ std::ffi::OsStr);
+impl_as_ref!(std::ffi::OsString);
+
+// Cow<'_, str> does not implement AsRef<OsStr> and this is unlikely to be fixed
+// for backward compatibility reasons. Otherwise we'd use a blanket impl.
+impl Quotable for Cow<'_, str> {
+    fn quote(&self) -> Quoted<'_> {
+        // I suspect there's a better way to do this but I haven't found one
+        Quoted::new(<Cow<'_, str> as Borrow<str>>::borrow(self).as_ref())
```
Comment (miDeb) [inline src/uucore/src/lib/mods/display.rs:83]: FWIW, `self.as_ref().as_ref()` seems to work.

File: src/uucore/src/lib/mods/display.rs [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+                // Unsafe characters according to:
+                // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_02_03
+                b'"' | b'`' | b'$' | b'\\' => is_double_safe = false,
+                _ => (),
+            }
+            if !requires_quote && SPECIAL_SHELL_CHARS.contains(&ch) {
+                requires_quote = true;
+            }
+        }
+        let text = match from_utf8(text) {
```
Comment (miDeb) [inline src/uucore/src/lib/mods/display.rs:159]: I think we can move the conversion to a `&str` in front of the `for` loop above.

File: src/uucore/src/lib/mods/display.rs [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+            Err(_) => return write_escaped(f, text, self.escape_control),
+            Ok(text) => text,
+        };
+        if !requires_quote && text.find(char::is_whitespace).is_some() {
+            requires_quote = true;
```
Comment (miDeb) [inline src/uucore/src/lib/mods/display.rs:164]: Could we move the check for whitespace chars into the for loop?

**Comment by blyxxyz** [comment]
I thought about it a little more and now I agree with @tertsdiepraam. This setup isn't good enough for `ls` in the long term. For example, GNU `ls` lines up its results depending on whether they're quoted:
```
$ ls
 baz
'foo bar'
```
You could do that by checking for a string prefix, but at that point you're fighting against the API, and there are probably other things `ls` would want to do that I'm not aware of.

I think it would be less of a burden to outright copy relevant bits of the implementation into `quoting_style`.

I'll remove the `ls` integration. (Would it be ok to keep d3454ff2b55fc00879db67ccb64f81f2acafa3f3? That seems harmless.)

**Comment by blyxxyz** [comment]
The ls integration is now gone. I kept a bugfix for terminal detection (c737887efe326f05b84e601da673f886793fc839) and some changes to make it `OsStr`-oriented (778b8fcebbb6aa38e2f7ef694f432f3f8825dced, 1838ac53488a2585db305893e73c060ecdbc6532, can remove if @tertsdiepraam prefers).

I removed the quoting configuration to keep things simple but the flexible version is still [in the history](https://github.com/uutils/coreutils/blob/b1ec1e33f236032b1363560ca7f0d34862356df7/src/uucore/src/lib/mods/display.rs) for future reference.

**Comment by tertsdiepraam** [comment]
> can remove if @tertsdiepraam prefers

No need! I like those changes!

**Comment by blyxxyz** [comment]
I found a use for conditional quoting after all, so I added it back in:
```
$ wc notfound 'not found'
wc: notfound: No such file or directory
wc: 'not found': No such file or directory
```

File: src/uucore/src/lib/mods/display.rs [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+    /// #[macro_use]
+    /// extern crate uucore;
+    /// use std::path::Path;
+    /// use uucore::display::Quotable;
+    ///
+    /// let foo = Path::new("foo/bar.baz");
+    /// let bar = Path::new("foo bar");
+    ///
+    /// show_error!("{}: Not found", foo); // Prints "util: foo/bar.baz: Not found"
+    /// show_error!("{}: Not found", bar); // Prints "util: 'foo bar': Not found"
```
Comment (miDeb) [inline src/uucore/src/lib/mods/display.rs:72]: This doc test doesn't pass.

**Review by miDeb — APPROVED** [review]
Thanks for figuring this out! It's great that we can achieve a better and more understandable quoting style than gnu.  If @tertsdiepraam agrees I think we can merge this after the doc test is fixed.

I assume we want to replace all occurences of `path.display` in `uutils` with `.quote` or `.maybe_quote`? If so, I think we should use the [`disallowed_method` lint](https://rust-lang.github.io/rust-clippy/master/#disallowed_method) to forbid the use of `Path::display()` in the codebase.

**Review by tertsdiepraam — APPROVED** [review]

**Comment by blyxxyz** [comment]
> If so, I think we should use the [`disallowed_method` lint](https://rust-lang.github.io/rust-clippy/master/#disallowed_method) to forbid the use of `Path::display()` in the codebase.

I like this idea, but it doesn't seem trivial. It's "allow" by default, and there doesn't seem to be a way to change a lint's level for the whole workspace (at least not in `clippy.toml`, that idea was rejected). So we'd have to add `#![warn(clippy::disallowed_method)]` to a hundred crates and make sure it ends up in any new ones.

Unless this is within the power of `uucore_procs::main!()` to add? Putting it in the obvious place doesn't work:
```
error: an inner attribute is not permitted in this context
 --> src/uu/mkfifo/src/main.rs:1:1
  |
1 | uucore_procs::main!(uu_mkfifo);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: inner attributes, like `#![no_std]`, annotate the item enclosing them, and are usually found at the beginning of source files. Outer attributes, like `#[test]`, annotate the item following them.
```

Either way, I'll do a follow-up PR changing the current instances by ripgrepping for `.display()` and `'{}'`.

**Comment by tertsdiepraam** [comment]
I don't think applying the `disallowed_method` lint to all crates is correct. For example, `ls --quoting-style=literal` would ideally use `Path::Display` (even though it converts to `String` at the moment).

**Comment by blyxxyz** [comment]
Ideally that'd use something like `display::println_verbatim`. Perhaps not that API specifically, maybe it should write to a byte buffer that is later `write_all()`'d, but `Path::display` is lossy and that's not ideal.

-----

One thing I don't fully understand yet is how it should be handled on Windows, but writing lossily there would work for now (and bad filenames are much rarer there than on Unix). As far as I can tell it transcodes from UTF-8 to UTF-16 when writing to stdout/stderr (and errors if the output isn't UTF-8) but writes bytes verbatim when writing to a file.

Even piped data has to be 16-bit, you can't pipe binary data (at least in the Linux build of Powershell, even without involving Rust, and I assume it extends to Powershell on Windows). So when printing/piping we'd want to manually write a 16-bit buffer through `winapi`, and when writing to a file it would be better to do lossy UTF-8 because all existing code probably writes UTF-8 given that it's Rust's default.

Not sure about a good API for that.

-----

I'm already going through all instances of `'{}'`/`{}:`/`.display()`/`.to_string_lossy()` so I'll see what else turns up.

**Comment by tertsdiepraam** [comment]
>  but Path::display is lossy and that's not ideal.

Good point!

**Pull request merged**